File: cpp/quantum/error_correction.cpp
#include "error_correction.hpp"

#include <stdexcept>

namespace qminiwasm::quantum {

// ============================================================================
// [[5,1,3]] Code Implementation
// ============================================================================

QutritTableau Code513::encode(int state) const {
  QutritTableau tab(N_PHYSICAL);

  // Encoding circuit for [[5,1,3]] code
  // Step 1: Create entanglement
  tab.apply_h3(0);
  tab.apply_cz3(0, 1);
  tab.apply_cz3(0, 2);
  tab.apply_cz3(0, 3);
  tab.apply_cz3(0, 4);

  // Step 2: Apply stabilizer structure
  tab.apply_h3(1);
  tab.apply_cz3(1, 2);
  tab.apply_h3(2);
  tab.apply_cz3(2, 3);
  tab.apply_h3(3);
  tab.apply_cz3(3, 4);

  // Encode the logical state
  if (state == 1) {
    tab.apply_x3(0);
  } else if (state == 2) {
    tab.apply_x3(0, 2);
  }

  return tab;
}

std::array<int, 4> Code513::extract_syndrome(const QutritTableau& tab) const {
  std::array<int, 4> syndrome{};
  const auto& x = tab.x();
  const auto& z = tab.z();

  for (int i = 0; i < 4; ++i) {
    int s = 0;
    for (int j = 0; j < N_PHYSICAL; ++j) {
      // Symplectic inner product: H_x[i,j] * Z[j] + H_z[i,j] * X[j]
      s += X_CHECK[i][j] * z[j * N_PHYSICAL + j];
      s += Z_CHECK[i][j] * x[j * N_PHYSICAL + j];
    }
    syndrome[i] = ((s % 3) + 3) % 3;
  }

  return syndrome;
}

std::optional<std::pair<int, int>> Code513::identify_error(
    const std::array<int, 4>& syndrome) const {
  // Check for trivial syndrome
  if (syndrome[0] == 0 && syndrome[1] == 0 && syndrome[2] == 0 && syndrome[3] == 0) {
    return std::nullopt;
  }

  // Syndrome lookup table for [[5,1,3]] code
  // Maps syndrome to (qutrit, error_type)
  // error_type: 0=X error, 1=Z error, 2=Y error
  static const std::array<std::pair<int, int>, 81> syndrome_map = []() {
    std::array<std::pair<int, int>, 81> map{};
    // Initialize all to invalid
    for (auto& p : map) {
      p = {-1, -1};
    }

    // X errors on each qutrit
    map[1 * 27 + 0 * 9 + 0 * 3 + 2] = {0, 0};  // (1,0,0,2) -> X on qubit 0
    map[2 * 27 + 0 * 9 + 0 * 3 + 1] = {0, 0};  // (2,0,0,1) -> X² on qubit 0
    map[0 * 27 + 1 * 9 + 0 * 3 + 2] = {1, 0};  // (0,1,0,2) -> X on qubit 1
    map[0 * 27 + 2 * 9 + 0 * 3 + 1] = {1, 0};  // (0,2,0,1) -> X² on qubit 1
    map[0 * 27 + 0 * 9 + 1 * 3 + 2] = {2, 0};  // (0,0,1,2) -> X on qubit 2
    map[0 * 27 + 0 * 9 + 2 * 3 + 1] = {2, 0};  // (0,0,2,1) -> X² on qubit 2
    map[2 * 27 + 0 * 9 + 0 * 3 + 0] = {3, 0};  // (2,0,0,0) -> X on qubit 3
    map[1 * 27 + 0 * 9 + 0 * 3 + 0] = {3, 0};  // (1,0,0,0) -> X² on qubit 3
    map[0 * 27 + 2 * 9 + 0 * 3 + 0] = {4, 0};  // (0,2,0,0) -> X on qubit 4
    map[0 * 27 + 1 * 9 + 0 * 3 + 0] = {4, 0};  // (0,1,0,0) -> X² on qubit 4

    // Z errors on each qutrit
    map[0 * 27 + 2 * 9 + 2 * 3 + 1] = {0, 1};  // (0,2,2,1) -> Z on qubit 0
    map[0 * 27 + 1 * 9 + 1 * 3 + 2] = {0, 1};  // (0,1,1,2) -> Z² on qubit 0
    map[1 * 27 + 0 * 9 + 2 * 3 + 2] = {1, 1};  // (1,0,2,2) -> Z on qubit 1
    map[2 * 27 + 0 * 9 + 1 * 3 + 1] = {1, 1};  // (2,0,1,1) -> Z² on qubit 1
    map[2 * 27 + 1 * 9 + 0 * 3 + 2] = {2, 1};  // (2,1,0,2) -> Z on qubit 2
    map[1 * 27 + 2 * 9 + 0 * 3 + 1] = {2, 1};  // (1,2,0,1) -> Z² on qubit 2
    map[2 * 27 + 2 * 9 + 1 * 3 + 0] = {3, 1};  // (2,2,1,0) -> Z on qubit 3
    map[1 * 27 + 1 * 9 + 2 * 3 + 0] = {3, 1};  // (1,1,2,0) -> Z² on qubit 3
    map[1 * 27 + 2 * 9 + 2 * 3 + 2] = {4, 1};  // (1,2,2,2) -> Z on qubit 4
    map[2 * 27 + 1 * 9 + 1 * 3 + 1] = {4, 1};  // (2,1,1,1) -> Z² on qubit 4

    return map;
  }();

  int idx = syndrome[0] * 27 + syndrome[1] * 9 + syndrome[2] * 3 + syndrome[3];
  if (idx >= 0 && idx < 81) {
    auto result = syndrome_map[idx];
    if (result.first >= 0) {
      return result;
    }
  }

  // Unknown syndrome (possibly Y error or multiple errors)
  return std::nullopt;
}

void Code513::correct_error(QutritTableau& tab, int qutrit, int error_type) const {
  if (error_type == 0) {
    // Apply X⁻¹ = X² to correct X error
    tab.apply_x3(qutrit, 2);
  } else if (error_type == 1) {
    // Apply Z⁻¹ = Z² to correct Z error
    tab.apply_z3(qutrit, 2);
  } else if (error_type == 2) {
    // Y = XZ, so apply Y⁻¹ = Z⁻¹X⁻¹ = Z²X²
    tab.apply_z3(qutrit, 2);
    tab.apply_x3(qutrit, 2);
  }
}

// ============================================================================
// [[3,1,2]] Code Implementation
// ============================================================================

QutritTableau Code312::encode(int state) const {
  QutritTableau tab(N_PHYSICAL);

  // Encoding circuit for [[3,1,2]] code
  tab.apply_h3(0);
  tab.apply_cz3(0, 1);
  tab.apply_cz3(0, 2);

  // Encode the logical state
  if (state == 1) {
    tab.apply_x3(0);
  } else if (state == 2) {
    tab.apply_x3(0, 2);
  }

  return tab;
}

std::pair<int, int> Code312::extract_syndrome(const QutritTableau& tab) const {
  const auto& x = tab.x();
  const auto& z = tab.z();

  int x_syn = 0;
  int z_syn = 0;

  for (int j = 0; j < N_PHYSICAL; ++j) {
    x_syn += x[j * N_PHYSICAL + j];
    z_syn += z[j * N_PHYSICAL + j];
  }

  return {((x_syn % 3) + 3) % 3, ((z_syn % 3) + 3) % 3};
}

bool Code312::is_valid(const std::pair<int, int>& syndrome) const {
  return syndrome.first == 0 && syndrome.second == 0;
}

// ============================================================================
// Convenience Functions
// ============================================================================

QutritTableau encode_513(int state) {
  return Code513().encode(state);
}

QutritTableau encode_312(int state) {
  return Code312().encode(state);
}

bool correct_single_error_513(QutritTableau& tab) {
  Code513 code;
  auto syndrome = code.extract_syndrome(tab);
  auto error = code.identify_error(syndrome);

  if (error.has_value()) {
    auto [qutrit, error_type] = error.value();
    code.correct_error(tab, qutrit, error_type);
    return true;
  }

  return false;
}

bool detect_error_312(const QutritTableau& tab) {
  Code312 code;
  auto syndrome = code.extract_syndrome(tab);
  return !code.is_valid(syndrome);
}

}  // namespace qminiwasm::quantum

File: qminiwasm/quantum/clifford/error_correction.py
"""Qutrit stabilizer error correction codes.

This module implements the [[5,1,3]] and [[3,1,2]] qutrit stabilizer codes for
error detection and correction in ternary neural networks. These codes protect
against quantization drift and physical memory bit-flips in edge environments.

The [[5,1,3]] code encodes 1 logical qutrit into 5 physical qutrits with code
distance 3, enabling correction of any single-qutrit error (X, Z, or Y type).

The [[3,1,2]] code encodes 1 logical qutrit into 3 physical qutrits with code
distance 2, providing erasure detection for less critical network layers.
"""

from __future__ import annotations

from typing import List, Optional, Tuple
import numpy as np

from qminiwasm.quantum.clifford.qutrit_tableau import QutritStabilizerTableau
from qminiwasm.quantum.clifford.qutrit_gates import (
    apply_h3, apply_s3, apply_cz3, apply_x3, apply_z3
)


class QutritErrorCode:
    """Base class for qutrit error correction codes."""

    def __init__(self, n_physical: int, k_logical: int, distance: int):
        """Initialize error code.

        Args:
            n_physical: Number of physical qutrits
            k_logical: Number of logical qutrits
            distance: Code distance (minimum weight of undetectable error)
        """
        self.n = n_physical
        self.k = k_logical
        self.d = distance


class Code513(QutritErrorCode):
    """[[5,1,3]] qutrit stabilizer code.

    Encodes 1 logical qutrit into 5 physical qutrits.
    Code distance 3: corrects any single-qutrit error.

    Stabilizer generators (cyclic structure):
    g₁ = X₁ Z₂ Z₃ X₄ I₅
    g₂ = I₁ X₂ Z₃ Z₄ X₅
    g₃ = X₁ I₂ X₃ Z₄ Z₅
    g₄ = Z₁ X₂ I₃ X₄ Z₅
    """

    def __init__(self):
        super().__init__(n_physical=5, k_logical=1, distance=3)
        # Stabilizer generator check matrix [X | Z] over GF(3)
        # Each row is a stabilizer generator
        self._build_check_matrix()

    def _build_check_matrix(self) -> None:
        """Build the check matrix for syndrome extraction."""
        # X components of stabilizers
        self.x_check = np.array([
            [1, 0, 0, 1, 0],  # g₁: X₁ Z₂ Z₃ X₄ I₅
            [0, 1, 0, 0, 1],  # g₂: I₁ X₂ Z₃ Z₄ X₅
            [1, 0, 1, 0, 0],  # g₃: X₁ I₂ X₃ Z₄ Z₅
            [0, 1, 0, 1, 0],  # g₄: Z₁ X₂ I₃ X₄ Z₅
        ], dtype=np.int32) % 3

        # Z components of stabilizers
        self.z_check = np.array([
            [0, 1, 1, 0, 0],  # g₁: X₁ Z₂ Z₃ X₄ I₅
            [0, 0, 1, 1, 0],  # g₂: I₁ X₂ Z₃ Z₄ X₅
            [0, 0, 0, 1, 1],  # g₃: X₁ I₂ X₃ Z₄ Z₅
            [1, 0, 0, 0, 1],  # g₄: Z₁ X₂ I₃ X₄ Z₅
        ], dtype=np.int32) % 3

    def encode(self, state: int = 0) -> QutritStabilizerTableau:
        """Encode a logical qutrit into the [[5,1,3]] code.

        Args:
            state: Logical state to encode (0, 1, or 2)

        Returns:
            QutritStabilizerTableau representing the encoded state
        """
        tab = QutritStabilizerTableau(self.n)

        # Initialize to |0⟩^5
        # Apply encoding circuit to create logical |0⟩ state
        # This creates the stabilizer state corresponding to logical |0⟩

        # Encoding circuit for [[5,1,3]] code
        # Step 1: Create entanglement
        apply_h3(tab, 0)
        apply_cz3(tab, 0, 1)
        apply_cz3(tab, 0, 2)
        apply_cz3(tab, 0, 3)
        apply_cz3(tab, 0, 4)

        # Step 2: Apply stabilizer structure
        apply_h3(tab, 1)
        apply_cz3(tab, 1, 2)
        apply_h3(tab, 2)
        apply_cz3(tab, 2, 3)
        apply_h3(tab, 3)
        apply_cz3(tab, 3, 4)

        # Encode the logical state
        if state == 1:
            apply_x3(tab, 0)
        elif state == 2:
            apply_x3(tab, 0, power=2)

        return tab

    def extract_syndrome(self, tab: QutritStabilizerTableau) -> Tuple[int, ...]:
        """Extract error syndrome from the tableau.

        The syndrome is computed via the symplectic inner product over GF(3):
        s_i = (H_x · e_z + H_z · e_x) mod 3

        where e_x, e_z are the error X/Z components.

        Args:
            tab: QutritStabilizerTableau to analyze

        Returns:
            Tuple of syndrome values (0, 1, or 2) for each stabilizer
        """
        syndrome = []
        for i in range(4):  # 4 stabilizer generators
            # Compute symplectic inner product
            s = 0
            for j in range(self.n):
                # H_x[i,j] * Z[j] + H_z[i,j] * X[j]
                s += self.x_check[i, j] * int(tab.z[j, j])
                s += self.z_check[i, j] * int(tab.x[j, j])
            syndrome.append(s % 3)
        return tuple(syndrome)

    def identify_error(self, syndrome: Tuple[int, ...]) -> Optional[Tuple[int, int]]:
        """Identify error location and type from syndrome.

        Args:
            syndrome: Error syndrome from extract_syndrome

        Returns:
            Tuple of (qutrit_index, error_type) where:
            - qutrit_index: 0-4 for physical qutrit
            - error_type: 0=X error, 1=Z error, 2=Y error
            Returns None if syndrome is trivial (no error)
        """
        if all(s == 0 for s in syndrome):
            return None

        # Syndrome lookup table for [[5,1,3]] code
        # Maps syndrome to (qutrit, error_type)
        # This is computed from the check matrix structure
        syndrome_map = {
            (1, 0, 0, 2): (0, 0),  # X error on qubit 0
            (2, 0, 0, 1): (0, 0),  # X² error on qubit 0
            (0, 1, 0, 2): (1, 0),  # X error on qubit 1
            (0, 2, 0, 1): (1, 0),  # X² error on qubit 1
            (0, 0, 1, 2): (2, 0),  # X error on qubit 2
            (0, 0, 2, 1): (2, 0),  # X² error on qubit 2
            (2, 0, 0, 0): (3, 0),  # X error on qubit 3
            (1, 0, 0, 0): (3, 0),  # X² error on qubit 3
            (0, 2, 0, 0): (4, 0),  # X error on qubit 4
            (0, 1, 0, 0): (4, 0),  # X² error on qubit 4
            (0, 2, 2, 1): (0, 1),  # Z error on qubit 0
            (0, 1, 1, 2): (0, 1),  # Z² error on qubit 0
            (1, 0, 2, 2): (1, 1),  # Z error on qubit 1
            (2, 0, 1, 1): (1, 1),  # Z² error on qubit 1
            (2, 1, 0, 2): (2, 1),  # Z error on qubit 2
            (1, 2, 0, 1): (2, 1),  # Z² error on qubit 2
            (2, 2, 1, 0): (3, 1),  # Z error on qubit 3
            (1, 1, 2, 0): (3, 1),  # Z² error on qubit 3
            (1, 2, 2, 2): (4, 1),  # Z error on qubit 4
            (2, 1, 1, 1): (4, 1),  # Z² error on qubit 4
        }

        # Y errors have syndromes that are sums of X and Z syndromes
        # For simplicity, we can detect them but may need combination
        return syndrome_map.get(syndrome)

    def correct_error(
        self,
        tab: QutritStabilizerTableau,
        qutrit: int,
        error_type: int
    ) -> None:
        """Apply correction for identified error.

        Args:
            tab: QutritStabilizerTableau to correct
            qutrit: Qutrit index with error
            error_type: 0=X error, 1=Z error, 2=Y error
        """
        if error_type == 0:
            # Apply X⁻¹ = X² to correct X error
            apply_x3(tab, qutrit, power=2)
        elif error_type == 1:
            # Apply Z⁻¹ = Z² to correct Z error
            apply_z3(tab, qutrit, power=2)
        elif error_type == 2:
            # Y = XZ, so apply Y⁻¹ = Z⁻¹X⁻¹ = Z²X²
            apply_z3(tab, qutrit, power=2)
            apply_x3(tab, qutrit, power=2)


class Code312(QutritErrorCode):
    """[[3,1,2]] qutrit stabilizer code.

    Encodes 1 logical qutrit into 3 physical qutrits.
    Code distance 2: detects single erasures (but cannot correct).

    Stabilizer generators:
    g₁ = X₁ X₂ X₃
    g₂ = Z₁ Z₂ Z₃
    """

    def __init__(self):
        super().__init__(n_physical=3, k_logical=1, distance=2)
        self._build_check_matrix()

    def _build_check_matrix(self) -> None:
        """Build the check matrix for syndrome extraction."""
        self.x_check = np.array([
            [1, 1, 1],  # g₁: X₁ X₂ X₃
            [0, 0, 0],  # g₂: Z₁ Z₂ Z₃ (no X component)
        ], dtype=np.int32) % 3

        self.z_check = np.array([
            [0, 0, 0],  # g₁: X₁ X₂ X₃ (no Z component)
            [1, 1, 1],  # g₂: Z₁ Z₂ Z₃
        ], dtype=np.int32) % 3

    def encode(self, state: int = 0) -> QutritStabilizerTableau:
        """Encode a logical qutrit into the [[3,1,2]] code.

        Args:
            state: Logical state to encode (0, 1, or 2)

        Returns:
            QutritStabilizerTableau representing the encoded state
        """
        tab = QutritStabilizerTableau(self.n)

        # Encoding circuit for [[3,1,2]] code
        apply_h3(tab, 0)
        apply_cz3(tab, 0, 1)
        apply_cz3(tab, 0, 2)

        # Encode the logical state
        if state == 1:
            apply_x3(tab, 0)
        elif state == 2:
            apply_x3(tab, 0, power=2)

        return tab

    def extract_syndrome(self, tab: QutritStabilizerTableau) -> Tuple[int, int]:
        """Extract error syndrome from the tableau.

        Args:
            tab: QutritStabilizerTableau to analyze

        Returns:
            Tuple of (x_syndrome, z_syndrome) values
        """
        x_syn = 0
        z_syn = 0
        for j in range(self.n):
            x_syn += int(tab.x[j, j])
            z_syn += int(tab.z[j, j])
        return (x_syn % 3, z_syn % 3)

    def is_valid(self, syndrome: Tuple[int, int]) -> bool:
        """Check if syndrome indicates a valid (error-free) state.

        Args:
            syndrome: Error syndrome

        Returns:
            True if no error detected
        """
        return syndrome == (0, 0)


# Convenience functions


def encode_513(state: int = 0) -> QutritStabilizerTableau:
    """Encode a logical qutrit using the [[5,1,3]] code.

    Args:
        state: Logical state (0, 1, or 2)

    Returns:
        Encoded QutritStabilizerTableau
    """
    code = Code513()
    return code.encode(state)


def encode_312(state: int = 0) -> QutritStabilizerTableau:
    """Encode a logical qutrit using the [[3,1,2]] code.

    Args:
        state: Logical state (0, 1, or 2)

    Returns:
        Encoded QutritStabilizerTableau
    """
    code = Code312()
    return code.encode(state)


def correct_single_error_513(tab: QutritStabilizerTableau) -> bool:
    """Detect and correct a single-qutrit error in [[5,1,3]] code.

    Args:
        tab: QutritStabilizerTableau to correct (modified in place)

    Returns:
        True if an error was detected and corrected
    """
    code = Code513()
    syndrome = code.extract_syndrome(tab)
    error = code.identify_error(syndrome)
    if error is not None:
        qutrit, error_type = error
        code.correct_error(tab, qutrit, error_type)
        return True
    return False


def detect_error_312(tab: QutritStabilizerTableau) -> bool:
    """Detect an error in [[3,1,2]] code.

    Args:
        tab: QutritStabilizerTableau to check

    Returns:
        True if an error was detected (but not corrected)
    """
    code = Code312()
    syndrome = code.extract_syndrome(tab)
    return not code.is_valid(syndrome)